lib/Borel/Types/UOM.hs

```diff
@@ -0,0 +1,225 @@
+        pretty Byte      = "B"
+        pretty Instance  = "instance"
+        pretty IPAddress = "ip-address"
+        pretty CPU       = "cpu"
+        pretty VCPU      = "vcpu-allocation"
+        parse "s"               = Just Second
+        parse "h"               = Just Hour
+        parse "B"               = Just Byte
+        parse "instance"        = Just Instance
+        parse "ip-address"      = Just IPAddress
+        parse "cpu"             = Just CPU
+        parse "vcpu-allocation" = Just VCPU
+        parse _                 = Nothing
+
+pPrefixUOM :: Prism' Text Prefix
+pPrefixUOM = prism' pretty parse
+  where pretty Base = ""
+        pretty Giga = "G"
+        pretty Nano = "n"
+        pretty Mebi = "Mi"
+        pretty Mega = "M"
+        parse ""   = Just Base
+        parse "G"  = Just Giga
+        parse "n"  = Just Nano
+        parse "Mi" = Just Mebi
+        parse "M"  = Just Mega
+        parse _    = Nothing
+
+pUOM :: Prism' Text UOM
+pUOM = prism' pretty parse
+  where dash = "-"
+        pretty (u `Times` v) = (u ^. re pUOM) <> dash <> (v ^. re pUOM)
+        pretty (UOM p b)     = (p ^. re pPrefixUOM) <> (b ^. re pBaseUOM)
+
+        -- treat UOM @Times@ as left-associative.
+        parse = hush . AT.parseOnly (parser <* AT.endOfInput)
+        parser = do
+          uoms <- puom `AT.sepBy` AT.string dash
+          case uoms of []     -> mzero
+                       -- construct left-associative UOMs
+                       (u:us) -> return $ foldl Times u us
+        puom = do
+          pre  <- AT.option (Just Base) ppre
+          base <- pbase
+          case (pre, base) of (Just p, Just b)  -> return $ UOM p b
+                              _                 -> mzero
+        ppre   =  (preview pPrefixUOM <$> AT.string "G")
+              <|> (preview pPrefixUOM <$> AT.string "n")
+              <|> (preview pPrefixUOM <$> AT.string "Mi")
+              <|> (preview pPrefixUOM <$> AT.string "M")
+        pbase =   (preview pBaseUOM <$> AT.string "s")
+              <|> (preview pBaseUOM <$> AT.string "h")
+              <|> (preview pBaseUOM <$> AT.string "B")
+              <|> (preview pBaseUOM <$> AT.string "instance")
+              <|> (preview pBaseUOM <$> AT.string "ip-address")
+              <|> (preview pBaseUOM <$> AT.string "cpu")
+              <|> (preview pBaseUOM <$> AT.string "vcpu-allocation")
+
+instance Show BaseUOM where
+  show = T.unpack . review pBaseUOM
+
+instance Show Prefix where
+  show = T.unpack . review pPrefixUOM
+
+instance Show UOM where
+  show = T.unpack . review pUOM
+
+instance Read UOM where
+  readsPrec _ (T.pack -> x) = maybe [] (pure . (,"")) $ x ^? pUOM
+
+instance FromJSON UOM where
+  parseJSON (String t) = maybe mzero return $ t ^? pUOM
+  parseJSON _          = mzero
+
+instance ToJSON UOM where
+  toJSON x = String $ x ^. re pUOM
+
+
+-- Convenient conversions
+
+nanosecToSec :: (UOM, Word64) -> (UOM, Word64)
+nanosecToSec (u, v)
```

Comment:
Okay, that makes more sense. Given how much explaining it took for me to get that, I return to my first remark of "maybe a comment would be useful here".